fix charuco matchImagePoints

AleksandrPanov opened the pull request:
The problem was in python bindings in `void matchImagePoints(InputArrayOfArrays detectedCorners, InputArray detectedIds, OutputArray objPoints, OutputArray imgPoints)`. 

`detectedCorners` may include aruco marker corners as `vector<vector<Point2f>>` or `vector<Mat>`. Any `Mat` in `vector<Mat>` must have 4 rows for 4 corners for one marker (one marker has 4 corners).

`detectedCorners` may include Charuco corners as `vector<Point2f>` or `Mat`. Python bindings add extra dimension to `detectedCorners` and therefore `vector<Mat>` case is additionally  processed for charuco corners.


Until the PR merge, you can using custom `matchImagePoints()`:

```
def matchImagePoints(charuco_board, charuco_corners, charuco_ids):
    objPoints = []
    imgPoints = []
    for i in range(0, len(charuco_ids)):
        index = charuco_ids[i]
        objPoints.append(charuco_board.getChessboardCorners()[index])
        # objPoints[-1][0][1] = charuco_board.getRightBottomCorner()[1] - objPoints[-1][0][1]  #  set old axis direction
        imgPoints.append(charuco_corners[i])
    return np.array(objPoints), np.array(imgPoints)
```



Use sample from https://github.com/opencv/opencv/pull/23139 to see problem.

args:
-w=5 -h=7 -sl=0.04 -ml=0.02 -d=10 -v=choriginal.jpg

With fix:
![image](https://user-images.githubusercontent.com/22337800/212574516-20a89a7f-6ea2-4d52-a4f6-ccb0ca7fd090.png)



`detectedCorners` includes charuco corners in this case. Python bindings add extra dimension to `detectedCorners` and therefore without fix `detectedCorners` processed as aruco marker corners and `matchImagePoints` returns bad values:

![image](https://user-images.githubusercontent.com/22337800/212574560-b9923895-b697-4220-a6ca-64747cd70bdb.png)


### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

asmorkalov commented on modules/objdetect/include/opencv2/objdetect/aruco_board.hpp:
```diff
@@ -142,6 +142,19 @@ class CV_EXPORTS_W_SIMPLE CharucoBoard : public Board {
     CV_WRAP float getMarkerLength() const;
 
+    /** @brief Given a board configuration and a set of detected markers, returns the corresponding
+     * image points and object points to call solvePnP()
```
sovePnP is not the only use case. Please mention solvePnP in see also list,

asmorkalov commented on modules/objdetect/include/opencv2/objdetect/aruco_board.hpp:
```diff
@@ -142,6 +142,19 @@ class CV_EXPORTS_W_SIMPLE CharucoBoard : public Board {
+     * @param objPoints Vector of vectors of board marker points in the board coordinate space.
+     * @param imgPoints Vector of vectors of the projections of board marker corner points.
+     */
+    CV_WRAP void matchImagePoints(InputArray detectedCharuco, InputArray detectedIds,
+                                  OutputArray objPoints, OutputArray imgPoints) const;
+
```
Why do you need dedicated method for it?

AleksandrPanov replied:
removed

AleksandrPanov commented on modules/objdetect/src/aruco/aruco_board.cpp:
```diff
@@ -369,20 +369,29 @@ void CharucoBoardImpl::calcNearestMarkerCorners() {
+    // Any Mat in vector<Mat> must have 4 rows for 4 corners for one marker.
     if (detectedCorners.kind() == _InputArray::STD_VECTOR_VECTOR ||
-        detectedCorners.isMatVector() || detectedCorners.isUMatVector())
+        (detectedCorners.isMatVector() && detectedCorners.total() > 0ull && detectedCorners.getMat(0).total() == 4ull))
```
fixed

AleksandrPanov commented on modules/objdetect/src/aruco/aruco_board.cpp:
```diff
@@ -369,20 +369,29 @@ void CharucoBoardImpl::calcNearestMarkerCorners() {
 
 void CharucoBoardImpl::matchImagePoints(InputArrayOfArrays detectedCorners, InputArray detectedIds,
-                                   OutputArray _objPoints, OutputArray imgPoints) const {
+                                        OutputArray _objPoints, OutputArray imgPoints) const {
```
class `CharucoBoardImpl` contains field `objPoints`
rename `_objPoints` to `outObjPoints`

asmorkalov commented on modules/objdetect/include/opencv2/objdetect/aruco_board.hpp:
```diff
@@ -58,14 +58,17 @@ class CV_EXPORTS_W_SIMPLE Board {
-     * For CharucoBoard class you can set list of charuco corners.
-     * @param detectedIds List of identifiers for each marker or list of charuco identifiers for each corner.
-     * For CharucoBoard class you can set list of charuco identifiers for each corner.
+     * Board/GridBoard support std::vector<std::vector<Point2f>> or std::vector<Mat> for marker corners.
```
For cv::Board and cv::GridBoard the method expects...
For cv::CharucoBoard the method expects std::vector<Point2f> or Mat with ChAruco corners (chess board corners matched with Aruco markers).

AleksandrPanov replied:
fixed

asmorkalov commented on modules/objdetect/src/aruco/charuco_detector.cpp:
```diff
@@ -130,7 +130,8 @@ struct CharucoDetector::CharucoDetectorImpl {
         Mat approximatedRvec, approximatedTvec;
         Mat objPoints, imgPoints; // object and image points for the solvePnP function
         // printf("before board.matchImagePoints(markerCorners, markerIds, objPoints, imgPoints);\n");
-        board.matchImagePoints(markerCorners, markerIds, objPoints, imgPoints);
+        Board simpleBoard(board.getObjPoints(), board.getDictionary(), board.getIds());
+        simpleBoard.matchImagePoints(markerCorners, markerIds, objPoints, imgPoints);
         // printf("after board.matchImagePoints(markerCorners, markerIds, objPoints, imgPoints);\n");
```
Please remove commented printf.

AleksandrPanov replied:
removed

asmorkalov reviewed (approved):
:+1: Looks good!

alalek commented on modules/objdetect/src/aruco/aruco_board.cpp:
```diff
@@ -368,25 +367,29 @@ void CharucoBoardImpl::calcNearestMarkerCorners() {
+        objPnts[i] = chessboardCorners[pointId];
+        if (detectedCharuco.isMatVector()) { // python binding case
+            imgPnts[i].x = detectedCharuco.getMat((int)i).at<float>(0);
+            imgPnts[i].y = detectedCharuco.getMat((int)i).at<float>(1);
```
> `.at<float>`

Missing type checks.

AleksandrPanov replied:
add Assert: `CV_Assert(... && detectedCharuco.depth() == CV_32F)`
add Assert: `CV_Assert((int)detectedCharucoMat.channels() == 2)` for C++ case
add Assert: `CV_Assert((int)detectedCharucoVecMat.front().total() * detectedCharucoVecMat.front().channels() == 2)` for python binding case

alalek commented on modules/objdetect/src/aruco/aruco_board.cpp:
```diff
@@ -368,25 +367,29 @@ void CharucoBoardImpl::calcNearestMarkerCorners() {
-        Mat(objPnts).copyTo(_objPoints);
-        Mat(imgPnts).copyTo(imgPoints);
+        else
+            imgPnts[i] = detectedCharuco.getMat().at<Point2f>((int)i);
```
> `.at<Point2f>`

Missing type checks.

AleksandrPanov replied:
add Assert: CV_Assert(... && detectedCharuco.depth() == CV_32F)
add Assert: CV_Assert((int)detectedCharucoMat.channels() == 2) for C++ case
add Assert: CV_Assert((int)detectedCharucoVecMat.front().total() * detectedCharucoVecMat.front().channels() == 2) for python binding case

alalek commented on modules/objdetect/src/aruco/aruco_board.cpp:
```diff
@@ -368,25 +367,29 @@ void CharucoBoardImpl::calcNearestMarkerCorners() {
+    vector<Point3f> objPnts(nDetected);
+    vector<Point2f> imgPnts(nDetected);
+    for(size_t i = 0ull; i < nDetected; i++) {
+        int pointId = detectedIds.getMat().at<int>((int)i);
```
> `.getMat()`

This is an expensive method in general. Do NOT call it repeatedly in a for loops.

AleksandrPanov replied:
remove `.getMat()` from loops

alalek reviewed (changes requested):
From what other part of OpenCV code did you bring that design pattern?
Looks like we have malformed API design - `InputArrayOfArrays` is misused here.

/cc @vpisarev 

---

Accessing elements without type checks must be avoided.

AleksandrPanov commented on modules/objdetect/src/aruco/aruco_board.cpp:
```diff
@@ -368,25 +367,29 @@ void CharucoBoardImpl::calcNearestMarkerCorners() {
-        Mat(objPnts).copyTo(_objPoints);
-        Mat(imgPnts).copyTo(imgPoints);
+        else
+            imgPnts[i] = detectedCharuco.getMat().at<Point2f>((int)i);
```
remove `.getMat()` from loops

AleksandrPanov commented on modules/objdetect/src/aruco/aruco_board.cpp:
```diff
@@ -368,25 +367,29 @@ void CharucoBoardImpl::calcNearestMarkerCorners() {
+    CV_Assert(detectedIds.total() == detectedCharuco.total());
+    CV_Assert(detectedIds.total() > 0ull);
+    // detectedCharuco includes charuco corners as vector<Point2f> or Mat.
+    // Python bindings could add extra dimension to detectedCharuco and therefore vector<Mat> case is additionally rocessed.
```
fixed typo

AleksandrPanov replied:
I suggest @asmorkalov to create new `CharucoBoard::matchImagePoints` with **InputArray** signature:

```
    CV_WRAP void matchImagePoints(InputArray detectedCharuco, InputArray detectedIds,
                                  OutputArray objPoints, OutputArray imgPoints) const;
```

But in this case, two Python functions are generated:

- matchImagePoints(**InputArray** detectedCharuco, ...)
- matchImagePoints(**InputArrayOfArrays** detectedCharuco, ...)

To avoid a potential problem @asmorkalov suggest use only default `Board::matchImagePoints` method

I don't know which way is better.

AleksandrPanov commented:
I fixed most of the issue (also updated docs, added cpp test). The issue of `InputArrayOfArrays` API design is discussed above in the conversation.

alalek commented on modules/objdetect/include/opencv2/objdetect/aruco_board.hpp:
```diff
@@ -58,14 +58,22 @@ class CV_EXPORTS_W_SIMPLE Board {
+     * @param objPoints Vector of marker points in the board coordinate space.
+     * For any Board class the method expects std::vector<cv::Point3f> objectPoints or cv::Mat
+     *
+     @param imgPoints Vector of marker points in the image coordinate space.
```
Which board implementation returns output `imgPoint` which is not the same to `detectedCorners`?

AleksandrPanov replied:
`imgPoint` could be reordered relative to `detectedCorners`

alalek commented on modules/objdetect/test/test_charucodetection.cpp:
```diff
@@ -656,4 +656,30 @@ TEST(Charuco, issue_14014)
+
+    vector<Point3f> objPoints;
+    vector<Point2f> imagePoints;
+    board.matchImagePoints(detectedCharucoCorners, detectedIds, objPoints, imagePoints);
+
+    ASSERT_EQ(detectedCharucoCorners.size(), objPoints.size());
+    ASSERT_EQ(detectedCharucoCorners.size(), imagePoints.size());
+
+    for (size_t i = 0ull; i < detectedCharucoCorners.size(); i++) {
+        EXPECT_EQ(detectedCharucoCorners[i], imagePoints[i]);
+        EXPECT_EQ(detectedCharucoCorners[i].x, objPoints[i].x);
+        EXPECT_EQ(detectedCharucoCorners[i].y, objPoints[i].y);
+    }
```
> detectedCharucoCorners
> objPoints

Test check that they are the same.
So why do we need such API? We should avoid redundancy.

AleksandrPanov replied:
They (`detectedCharucoCorners` and `objPoints`) were same only in this test

changed the value to make the test more realistic

AleksandrPanov replied:
also fixed checks:

```
    for (size_t i = 0ull; i < detectedCharucoCorners.size(); i++) {
        EXPECT_EQ(detectedCharucoCorners[i], imagePoints[i]);
        EXPECT_EQ(chessboardPoints[i].x, objPoints[i].x);
        EXPECT_EQ(chessboardPoints[i].y, objPoints[i].y);
    }
```

AleksandrPanov commented on modules/objdetect/src/aruco/aruco_board.cpp:
```diff
@@ -368,25 +374,40 @@ void CharucoBoardImpl::calcNearestMarkerCorners() {
+    CV_Assert(detectedCharuco.depth() == CV_32F);
+    // detectedCharuco includes charuco corners as vector<Point2f> or Mat.
+    // Python bindings could add extra dimension to detectedCharuco and therefore vector<Mat> case is additionally processed.
+    CV_Assert((detectedCharuco.isMat() || detectedCharuco.isVector() || detectedCharuco.isMatVector()) && detectedCharuco.depth() == CV_32F);
```
add UMatVector() support, was deleted by accident

AleksandrPanov commented on modules/objdetect/src/aruco/aruco_board.cpp:
```diff
@@ -368,25 +374,40 @@ void CharucoBoardImpl::calcNearestMarkerCorners() {
-        Mat(imgPnts).copyTo(imgPoints);
+void CharucoBoardImpl::matchImagePoints(InputArrayOfArrays detectedCharuco, InputArray detectedIds,
+                                        OutputArray outObjPoints, OutputArray outImgPoints) const {
+    CV_Assert(detectedIds.total() == detectedCharuco.total());
```
add:
`CV_CheckEQ(detectedIds.total(), detectedCharuco.total(), "Number of corners and ids must be equal");`

AleksandrPanov replied:
I have seen in most cases the use of `CV_Assert`. Is this a new standard? Now always add `CV_Check()` instead of `CV_Assert()`?

AleksandrPanov replied:
removed

AleksandrPanov commented on modules/objdetect/src/aruco/aruco_board.cpp:
```diff
@@ -368,25 +374,40 @@ void CharucoBoardImpl::calcNearestMarkerCorners() {
+        CV_Assert(pointId >= 0 && pointId < (int)chessboardCorners.size());
+        objPnts[i] = chessboardCorners[pointId];
+        if (detectedCharuco.isMatVector()) // python binding case
+            imgPnts[i] = detectedCharucoVecMat[i].ptr<Point2f>(0)[0];
```
added `CV_Assert((int)detectedCharucoVecMat[i].total() * detectedCharucoVecMat[i].channels() == 2);`

AleksandrPanov commented on modules/objdetect/src/aruco/aruco_board.cpp:
```diff
@@ -368,25 +374,40 @@ void CharucoBoardImpl::calcNearestMarkerCorners() {
+    Mat detectedCharucoMat, detectedIdsMat = detectedIds.getMat();
+    if (!detectedCharuco.isMatVector()) {
+        detectedCharucoMat = detectedCharuco.getMat();
+        CV_Assert((int)detectedCharucoMat.channels() == 2);
```
Is it necessary to do this? This expression `CV_Assert\([a-z, !,_]*.channels()` is used more often than checkVector
![image](https://user-images.githubusercontent.com/22337800/233490777-4d3f1b3f-6152-4caf-9f7b-930f916c2247.png)
![image](https://user-images.githubusercontent.com/22337800/233491179-22294c00-9ec5-4e19-8057-01cddface7fb.png)

AleksandrPanov commented on modules/objdetect/include/opencv2/objdetect/aruco_board.hpp:
```diff
@@ -58,14 +58,22 @@ class CV_EXPORTS_W_SIMPLE Board {
+     * @param objPoints Vector of marker points in the board coordinate space.
+     * For any Board class the method expects std::vector<cv::Point3f> objectPoints or cv::Mat
+     *
+     @param imgPoints Vector of marker points in the image coordinate space.
```
fixed

AleksandrPanov commented on modules/objdetect/src/aruco/aruco_board.cpp:
```diff
@@ -368,25 +374,40 @@ void CharucoBoardImpl::calcNearestMarkerCorners() {
+        int pointId = detectedIdsMat.at<int>((int)i);
+        CV_Assert(pointId >= 0 && pointId < (int)chessboardCorners.size());
+        objPnts[i] = chessboardCorners[pointId];
+        if (detectedCharuco.isMatVector()) // python binding case
```
removed

AleksandrPanov commented on modules/objdetect/include/opencv2/objdetect/aruco_board.hpp:
```diff
@@ -58,14 +58,22 @@ class CV_EXPORTS_W_SIMPLE Board {
+     *
+     *  @sa solvePnP
      */
     CV_WRAP void matchImagePoints(InputArrayOfArrays detectedCorners, InputArray detectedIds,
```
I suggest just fixing the bug for now. I will change the API in the following PRs.

AleksandrPanov commented on modules/objdetect/src/aruco/aruco_board.cpp:
```diff
@@ -368,25 +374,40 @@ void CharucoBoardImpl::calcNearestMarkerCorners() {
-        }
-        Mat(objPnts).copyTo(_objPoints);
-        Mat(imgPnts).copyTo(imgPoints);
+void CharucoBoardImpl::matchImagePoints(InputArrayOfArrays detectedCharuco, InputArray detectedIds,
```
I suggest just fixing the bug for now. I will change the API in the following PRs.

AleksandrPanov commented on modules/objdetect/misc/python/test/test_objdetect_aruco.py:
```diff
@@ -142,5 +142,33 @@ def test_charuco_detector(self):
+        aruco_dict = cv.aruco.getPredefinedDictionary(cv.aruco.DICT_4X4_50)
+        board_size = (3, 4)
+        board = cv.aruco.GridBoard(board_size, 5.0, 1.0, aruco_dict)
+        aruco_corners = np.array(board.getObjPoints())[:, :, :2]
```
I am creating pseudo-detected `aruco_corners` based on `objPoints`, `aruco_corners` uses 2d points (`Point2f`), but `objPoints` uses 3d points (`Point3f`) and we need 3D -> 2D conversion.

AleksandrPanov commented on modules/objdetect/include/opencv2/objdetect/aruco_board.hpp:
```diff
@@ -58,14 +58,22 @@ class CV_EXPORTS_W_SIMPLE Board {
+     @param imgPoints Vector of marker points in the image coordinate space.
+     * For any Board class the method expects std::vector<cv::Point2f> objectPoints or cv::Mat
+     *
+     *  @sa solvePnP
```
fixed
